File: deploy.php
<?php
/**
 * Kanji-Back Deployment Helper (PHP Version)
 * Upload this file to your server and access via browser
 * WARNING: Delete this file after deployment for security!
 */

set_time_limit(300); // 5 minutes timeout
ini_set('memory_limit', '256M');

class DeploymentHelper {
    private $uploadDir;
    private $logFile;
    
    public function __construct() {
        $this->uploadDir = __DIR__;
        $this->logFile = $this->uploadDir . '/deployment.log';
    }
    
    public function log($message) {
        $timestamp = date('Y-m-d H:i:s');
        $logEntry = "[$timestamp] $message\n";
        echo "<p>$message</p>";
        file_put_contents($this->logFile, $logEntry, FILE_APPEND);
    }
    
    public function extractArchive() {
        $files = glob($this->uploadDir . '/kanji-back-*.tar.gz');
        
        if (empty($files)) {
            $this->log("❌ No archive found. Upload kanji-back-YYYYMMDD-HHMMSS.tar.gz first");
            return false;
        }
        
        $archive = $files[0];
        $archiveName = basename($archive);
        
        $this->log("📦 Found archive: $archiveName");
        $this->log("⏳ Extracting...");
        
        // Check if tar/gzip is available
        if (!function_exists('shell_exec')) {
            $this->log("❌ shell_exec not available on this server");
            return $this->extractWithPhar($archive);
        }
        
        try {
            $output = shell_exec("cd " . escapeshellarg($this->uploadDir) . " && tar -xzf " . escapeshellarg($archive) . " 2>&1");
            
            if (file_exists($this->uploadDir . '/src')) {
                $this->log("✓ Archive extracted successfully");
                return true;
            } else {
                $this->log("❌ Extraction failed: " . ($output ?: "Unknown error"));
                return false;
            }
        } catch (Exception $e) {
            $this->log("❌ Error: " . $e->getMessage());
            return false;
        }
    }
    
    private function extractWithPhar($archive) {
        try {
            $this->log("📦 Using PHP Phar extractor...");
            $phar = new PharData($archive);
            $phar->extractTo($this->uploadDir);
            
            if (file_exists($this->uploadDir . '/src')) {
                $this->log("✓ Archive extracted with Phar");
                return true;
            }
        } catch (Exception $e) {
            $this->log("❌ Phar extraction failed: " . $e->getMessage());
        }
        return false;
    }
    
    public function runComposer() {
        $this->log("📦 Running Composer install...");
        
        if (!file_exists($this->uploadDir . '/composer.json')) {
            $this->log("❌ composer.json not found");
            return false;
        }
        
        // Check if composer exists
        $composerPath = trim(shell_exec('which composer 2>/dev/null') ?: shell_exec('where composer 2>nul'));
        
        if (empty($composerPath) && !file_exists('/usr/local/bin/composer')) {
            $this->log("⚠️  Composer not found. Trying with php...");
            $composerPath = 'php composer.phar';
        } else {
            $composerPath = $composerPath ?: 'composer';
        }
        
        try {
            $output = shell_exec("cd " . escapeshellarg($this->uploadDir) . " && " . $composerPath . " install --no-dev --optimize-autoloader 2>&1");
            
            if (strpos($output, 'Successfully installed') !== false || 
                strpos($output, 'Package operations') !== false ||
                strpos($output, 'No packages') !== false) {
                $this->log("✓ Composer install complete");
                return true;
            } else {
                $this->log("⚠️  Composer output: " . substr($output, 0, 500));
                return true; // Continue anyway
            }
        } catch (Exception $e) {
            $this->log("⚠️  Composer error (non-fatal): " . $e->getMessage());
            return true; // Continue anyway
        }
    }
    
    public function setPermissions() {
        $this->log("🔐 Setting permissions...");
        
        $dirs = [
            'tmp',
            'tmp/cache',
            'tmp/logs',
            'logs'
        ];
        
        foreach ($dirs as $dir) {
            $path = $this->uploadDir . '/' . $dir;
            if (is_dir($path)) {
                if (@chmod($path, 0775)) {
                    $this->log("✓ Permissions set for: $dir");
                } else {
                    $this->log("⚠️  Could not set permissions for: $dir (may need manual FTP fix)");
                }
            }
        }
        
        return true;
    }
    
    public function clearCache() {
        $this->log("🧹 Clearing cache...");
        
        $cacheDirs = [
            'tmp/cache',
            'tmp/logs'
        ];
        
        foreach ($cacheDirs as $dir) {
            $path = $this->uploadDir . '/' . $dir;
            if (is_dir($path)) {
                $files = glob($path . '/*');
                foreach ($files as $file) {
                    if (is_file($file)) {
                        @unlink($file);
                    }
                }
                $this->log("✓ Cleared: $dir");
            }
        }
        
        return true;
    }
    
    public function testDeployment() {
        $this->log("🧪 Testing deployment...");
        
        $checks = [
            'src/Application.php' => 'Application core',
            'config/app.php' => 'Configuration',
            'webroot/index.php' => 'Web root',
            'vendor/autoload.php' => 'Composer dependencies'
        ];
        
        foreach ($checks as $file => $desc) {
            $path = $this->uploadDir . '/' . $file;
            if (file_exists($path)) {
                $this->log("✓ $desc found");
            } else {
                $this->log("❌ $desc NOT found");
            }
        }
        
        return true;
    }
    
    public function cleanupArchive() {
        $this->log("🧹 Cleaning up...");
        
        $files = glob($this->uploadDir . '/kanji-back-*.tar.gz');
        foreach ($files as $file) {
            if (@unlink($file)) {
                $this->log("✓ Deleted: " . basename($file));
            }
        }
        
        return true;
    }
}

// HTML/CSS
$html = <<<'HTML'
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Kanji-Back Deployment</title>
    <style>
        * { margin: 0; padding: 0; box-sizing: border-box; }
        body {
            font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Oxygen, Ubuntu, Cantarell, sans-serif;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            min-height: 100vh;
            display: flex;
            align-items: center;
            justify-content: center;
            padding: 20px;
        }
        .container {
            background: white;
            border-radius: 10px;
            box-shadow: 0 20px 60px rgba(0,0,0,0.3);
            max-width: 600px;
            width: 100%;
            padding: 40px;
        }
        h1 {
            color: #333;
            margin-bottom: 10px;
            font-size: 28px;
        }
        .subtitle {
            color: #666;
            margin-bottom: 30px;
            font-size: 14px;
        }
        .steps {
            background: #f5f5f5;
            border-left: 4px solid #667eea;
            padding: 20px;
            margin: 20px 0;
            border-radius: 5px;
        }
        .step {
            margin: 10px 0;
            color: #555;
            font-size: 14px;
        }
        .step strong {
            color: #333;
        }
        button {
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
            border: none;
            padding: 12px 30px;
            border-radius: 5px;
            font-size: 16px;
            cursor: pointer;
            margin: 10px 5px 10px 0;
            transition: transform 0.2s;
        }
        button:hover {
            transform: translateY(-2px);
            box-shadow: 0 5px 15px rgba(102, 126, 234, 0.4);
        }
        button:disabled {
            opacity: 0.5;
            cursor: not-allowed;
            transform: none;
        }
        .output {
            background: #1e1e1e;
            color: #00ff00;
            padding: 15px;
            border-radius: 5px;
            font-family: 'Courier New', monospace;
            font-size: 12px;
            max-height: 300px;
            overflow-y: auto;
            margin: 20px 0;
            display: none;
        }
        .output.active {
            display: block;
        }
        p {
            margin: 5px 0;
        }
        .success { color: #00ff00; }
        .error { color: #ff4444; }
        .warning { color: #ffaa00; }
        .info { color: #44aaff; }
        .warning-box {
            background: #fff3cd;
            border: 1px solid #ffc107;
            color: #856404;
            padding: 15px;
            border-radius: 5px;
            margin: 20px 0;
            font-size: 13px;
        }
        .success-box {
            background: #d4edda;
            border: 1px solid #28a745;
            color: #155724;
            padding: 15px;
            border-radius: 5px;
            margin: 20px 0;
            font-size: 13px;
        }
        a {
            color: #667eea;
            text-decoration: none;
        }
        a:hover {
            text-decoration: underline;
        }
    </style>
</head>
<body>
    <div class="container">
        <h1>🚀 Kanji-Back Deployment</h1>
        <p class="subtitle">PHP Deployment Helper</p>
        
        <div class="warning-box">
            ⚠️ <strong>SECURITY WARNING:</strong> Delete this file (deploy.php) immediately after deployment!<br>
            Keep it on your server = security risk
        </div>
        
        <div class="steps">
            <div class="step"><strong>Step 1:</strong> Upload archive via FTP (kanji-back-YYYYMMDD-HHMMSS.tar.gz)</div>
            <div class="step"><strong>Step 2:</strong> Upload this file (deploy.php) via FTP</div>
            <div class="step"><strong>Step 3:</strong> Access deploy.php in your browser</div>
            <div class="step"><strong>Step 4:</strong> Click "Run Deployment" button below</div>
            <div class="step"><strong>Step 5:</strong> Delete deploy.php for security</div>
        </div>
        
        <form method="post">
            <button type="submit" name="action" value="extract">📦 Extract Archive</button>
            <button type="submit" name="action" value="full">🚀 Run Full Deployment</button>
            <button type="submit" name="action" value="test">🧪 Test Deployment</button>
            <button type="submit" name="action" value="cleanup">🧹 Cleanup</button>
        </form>
        
        <div class="output" id="output"></div>
    </div>
    
    <script>
        document.querySelectorAll('form').forEach(form => {
            form.addEventListener('submit', function() {
                document.getElementById('output').classList.add('active');
            });
        });
    </script>
</body>
</html>
HTML;

// Check if it's a POST request with action
if ($_SERVER['REQUEST_METHOD'] === 'POST' && isset($_POST['action'])) {
    $helper = new DeploymentHelper();
    
    echo $html;
    echo '<div class="output active">';
    
    $action = $_POST['action'];
    
    switch ($action) {
        case 'extract':
            $helper->extractArchive();
            break;
            
        case 'full':
            $helper->log("🚀 Starting full deployment...");
            if ($helper->extractArchive()) {
                $helper->runComposer();
                $helper->setPermissions();
                $helper->clearCache();
                $helper->testDeployment();
                echo '<div class="success-box" style="margin-top: 20px;">';
                echo '✅ <strong>Deployment Complete!</strong><br>';
                echo 'Your application is ready. Test it:<br>';
                echo '<a href="chmn?hanzi=%E5%92%B2" target="_blank">http://your-domain/kanji-back/chmn?hanzi=%E5%92%B2</a>';
                echo '</div>';
            }
            break;
            
        case 'test':
            $helper->testDeployment();
            break;
            
        case 'cleanup':
            $helper->cleanupArchive();
            break;
    }
    
    echo '</div>';
    return;
}

// Initial page load
echo $html;
?>

File: deploy.py
#!/usr/bin/env python3
"""
Kanji-Back Deployment Script
Usage: python deploy.py <environment> <remote-host> <remote-user> <remote-path>
Example: python deploy.py production server.com deploy /var/www/kanji-back
"""

import os
import sys
import subprocess
import shutil
import tempfile
import datetime
import tarfile

class Colors:
    RED = '\033[0;31m'
    GREEN = '\033[0;32m'
    YELLOW = '\033[1;33m'
    BLUE = '\033[0;34m'
    NC = '\033[0m'

def print_header(msg):
    print(f"{Colors.YELLOW}{'='*50}{Colors.NC}")
    print(f"{Colors.YELLOW}{msg}{Colors.NC}")
    print(f"{Colors.YELLOW}{'='*50}{Colors.NC}")

def print_success(msg):
    print(f"{Colors.GREEN}✓ {msg}{Colors.NC}")

def print_error(msg):
    print(f"{Colors.RED}✗ {msg}{Colors.NC}")

def print_info(msg):
    print(f"{Colors.BLUE}ℹ {msg}{Colors.NC}")

def check_requirements():
    """Check if required tools are available"""
    required_tools = {
        'git': 'Git version control',
        'ssh': 'SSH client',
        'scp': 'SCP file transfer',
        'composer': 'PHP Composer',
    }
    
    missing = []
    for tool, desc in required_tools.items():
        if shutil.which(tool) is None:
            missing.append(f"{tool} ({desc})")
    
    if missing:
        print_error("Required tools not found:")
        for tool in missing:
            print(f"  - {tool}")
        return False
    
    print_success("All required tools found")
    return True

def check_git_status():
    """Ensure git working directory is clean"""
    try:
        result = subprocess.run(['git', 'diff-index', '--quiet', 'HEAD', '--'], 
                              capture_output=True)
        if result.returncode != 0:
            print_error("Uncommitted changes detected")
            subprocess.run(['git', 'status'])
            return False
        print_success("Git status clean")
        return True
    except Exception as e:
        print_error(f"Git check failed: {e}")
        return False

def create_archive(project_root, temp_dir):
    """Create deployment archive excluding unnecessary files"""
    exclude_patterns = {
        'vendor/',
        '.git',
        '.gitignore',
        'node_modules',
        'tests/',
        '.env.local',
        '.env.example',
        '*.log',
        'tmp/cache/*',
        'tmp/logs/*',
        '.DS_Store',
        'deploy.sh',
        'deploy.py',
        '__pycache__',
        '*.pyc',
        '.vscode',
        '.idea',
    }
    
    timestamp = datetime.datetime.now().strftime('%Y%m%d-%H%M%S')
    archive_name = f"kanji-back-{timestamp}.tar.gz"
    archive_path = os.path.join(temp_dir, archive_name)
    
    print_info(f"Creating archive: {archive_name}")
    
    with tarfile.open(archive_path, "w:gz") as tar:
        for root, dirs, files in os.walk(project_root):
            # Modify dirs in-place to skip excluded directories
            dirs[:] = [d for d in dirs if d not in exclude_patterns and 
                      not any(os.path.relpath(os.path.join(root, d), project_root).startswith(ep.rstrip('/*')) 
                             for ep in exclude_patterns)]
            
            for file in files:
                file_path = os.path.join(root, file)
                arcname = os.path.relpath(file_path, project_root)
                
                # Skip excluded files
                if any(arcname.startswith(ep.rstrip('/*')) or arcname.endswith(ep.lstrip('*')) 
                      for ep in exclude_patterns):
                    continue
                
                try:
                    tar.add(file_path, arcname=arcname)
                except Exception as e:
                    print_error(f"Failed to add {file}: {e}")
    
    print_success(f"Archive created: {archive_path}")
    return archive_path, archive_name

def upload_archive(archive_path, archive_name, remote_user, remote_host):
    """Upload archive to remote server"""
    print_info(f"Uploading {archive_name} to {remote_user}@{remote_host}")
    
    try:
        cmd = ['scp', archive_path, f"{remote_user}@{remote_host}:/tmp/{archive_name}"]
        subprocess.run(cmd, check=True)
        print_success("Upload complete")
        return True
    except subprocess.CalledProcessError as e:
        print_error(f"Upload failed: {e}")
        return False

def deploy_on_server(remote_user, remote_host, remote_path, archive_name):
    """Execute deployment commands on remote server"""
    print_info("Deploying on remote server...")
    
    deploy_commands = f"""
set -e

ARCHIVE_NAME="{archive_name}"
REMOTE_PATH="{remote_path}"
ARCHIVE_LOCATION="/tmp/$ARCHIVE_NAME"

echo "Step 1: Creating backup..."
if [ -d "$REMOTE_PATH" ]; then
    BACKUP_PATH="${{REMOTE_PATH}}.backup.$(date +%Y%m%d-%H%M%S)"
    cp -r "$REMOTE_PATH" "$BACKUP_PATH"
    echo "Backup created at: $BACKUP_PATH"
else
    mkdir -p "$REMOTE_PATH"
fi

echo "Step 2: Extracting archive..."
cd "$REMOTE_PATH"
tar -xzf "$ARCHIVE_LOCATION"

echo "Step 3: Installing PHP dependencies (production only)..."
composer install --no-dev --optimize-autoloader --no-interaction

echo "Step 4: Setting permissions..."
chmod -R 755 ./bin 2>/dev/null || true
chmod -R 755 ./webroot 2>/dev/null || true
chmod -R 775 ./tmp 2>/dev/null || true
chmod -R 775 ./logs 2>/dev/null || true

echo "Step 5: Clearing cache..."
rm -rf ./tmp/cache/* 2>/dev/null || true
rm -rf ./tmp/logs/* 2>/dev/null || true

echo "Step 6: Cleaning up archive..."
rm -f "$ARCHIVE_LOCATION"

echo "Deployment complete!"
"""
    
    try:
        ssh_cmd = ['ssh', f"{remote_user}@{remote_host}", deploy_commands]
        subprocess.run(ssh_cmd, check=True)
        print_success("Server deployment complete")
        return True
    except subprocess.CalledProcessError as e:
        print_error(f"Remote deployment failed: {e}")
        return False

def main():
    if len(sys.argv) < 5:
        print_error("Invalid arguments")
        print(f"Usage: {sys.argv[0]} <environment> <remote-host> <remote-user> <remote-path>")
        print(f"Example: {sys.argv[0]} production server.com deploy /var/www/kanji-back")
        sys.exit(1)
    
    environment = sys.argv[1]
    remote_host = sys.argv[2]
    remote_user = sys.argv[3]
    remote_path = sys.argv[4]
    
    print_header("Kanji-Back Deployment Script")
    print(f"Environment: {environment}")
    print(f"Remote Host: {remote_host}")
    print(f"Remote User: {remote_user}")
    print(f"Remote Path: {remote_path}\n")
    
    # Get project root
    project_root = os.path.dirname(os.path.abspath(__file__))
    
    # Change to project root
    os.chdir(project_root)
    
    # Step 1: Check requirements
    print_info("[1/5] Checking requirements...")
    if not check_requirements():
        sys.exit(1)
    print()
    
    # Step 2: Check git status
    print_info("[2/5] Checking git status...")
    if not check_git_status():
        sys.exit(1)
    print()
    
    # Step 3: Create archive
    print_info("[3/5] Creating deployment archive...")
    temp_dir = tempfile.mkdtemp()
    try:
        archive_path, archive_name = create_archive(project_root, temp_dir)
        print()
        
        # Step 4: Upload archive
        print_info("[4/5] Uploading to server...")
        if not upload_archive(archive_path, archive_name, remote_user, remote_host):
            sys.exit(1)
        print()
        
        # Step 5: Deploy on server
        print_info("[5/5] Deploying on server...")
        if not deploy_on_server(remote_user, remote_host, remote_path, archive_name):
            sys.exit(1)
        print()
        
    finally:
        # Cleanup
        print_info("Cleaning up temporary files...")
        shutil.rmtree(temp_dir, ignore_errors=True)
        print_success("Cleanup complete\n")
    
    print_header("Deployment Successful!")
    print_success(f"Application deployed to: {remote_path}")

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        print("\nDeployment cancelled by user")
        sys.exit(1)
    except Exception as e:
        print_error(f"Unexpected error: {e}")
        sys.exit(1)
